viewer/Viewer.cpp

```diff
@@ -784,62 +989,141 @@ namespace Enki
 		{
-			QPoint diff = event->pos() - mouseGrabPos;
-			if (event->buttons() & Qt::LeftButton)
+			if (isMovableByPicking(selectedObject))
 			{
-				if (event->modifiers() & Qt::ShiftModifier)
-				{
-					camera.pos.rx() -= 0.5 * cos(-camera.yaw) * (double)diff.y() + 0.5 * sin(-camera.yaw) * (double)diff.x();
-					camera.pos.ry() -= 0.5 * sin(-camera.yaw) * -(double)diff.y() + 0.5 * cos(-camera.yaw) * (double)diff.x();
-				}
-				else
-				{
-					camera.yaw -= 0.01 * (double)diff.x();
-					camera.pitch = clamp(camera.pitch - 0.01 * (double)diff.y(), -M_PI / 2, M_PI / 2);
-				}
+				QPoint diff = event->pos() - mouseGrabPos;
+				selectedObject->angle -= 0.01 * (double)diff.x();
+				mouseGrabPos = event->pos();
 			}
-			else if (event->buttons() & Qt::RightButton)
+		}
+
+		// move selected object if it's movable by picking
+		else if (event->buttons() & Qt::LeftButton)
+		{
+			if (isMovableByPicking(selectedObject) && (event->pos() - mouseGrabPos).manhattanLength() > 10 )
 			{
-				if (event->modifiers() & Qt::ShiftModifier)
+				if(!trackballView)
 				{
-					camera.altitude += -(double)diff.y();
+					world->SkipPhysicsObjectsList.insert(selectedObject);
+
+					selectedObject->pos = Point(pointedPoint.x(),pointedPoint.y());
+					selectedObject->speed = Vector(0,0);
+					selectedObject->angSpeed = 0;
 				}
 				else
 				{
-					// TODO: zoom
+					if(!messageList.empty() && messageList.back().first == controlError1)
+						messageList.back().second = 100;
+					else sendMessage(controlError1, 100);
 				}
 			}
-			
+			else if((event->pos() - mouseGrabPos).manhattanLength() > 10 )
+			{
+				// TODO multiselection area feature
+				// cool but realy hard to implement with the actual picking system
+			}
+		}
+
+		// rotate camera
+		else if (event->buttons() & Qt::RightButton)
+		{
+			QPoint diff = event->pos() - mouseGrabPos;
+			camera.yaw -= 0.01 * (double)diff.x();
+
+			double delta = 0.01;
+			camera.pitch = clamp(camera.pitch - 0.01 * (double)diff.y(), -M_PI / 2 + delta, M_PI / 2 - delta);
+
 			mouseGrabPos = event->pos();
 		}
+
+		// translate camera
+		else if (event->buttons() & Qt::MidButton)
+		{
+			if(!trackballView)
+			{
+				QPoint diff = event->pos() - mouseGrabPos;
+				double sensibility = 0.03;
+				camera.pos.rx() += sensibility * (diff.x()*camera.left.x() + diff.y()*camera.up.x());
+				camera.pos.ry() += sensibility * (diff.x()*camera.left.y() + diff.y()*camera.up.y());
+				camera.altitude += sensibility * (diff.x()*camera.left.z() + diff.y()*camera.up.z());
+				mouseGrabPos = event->pos();
+			}
+			else
+			{
+				if(!messageList.empty() && messageList.back().first == controlError2)
+					messageList.back().second = 100;
+				else sendMessage(controlError2, 100);
+			}
+		}
 	}
 	
 	void ViewerWidget::wheelEvent(QWheelEvent * event)
 	{
-		/*if (event->modifiers() & Qt::ShiftModifier)
+		// zoom
+		if(trackballView)
```

Comment:
Enki's coding convention is to put a space between `if` and the opening parenthesis: `if (`.